// @unidraw/backend/src/server.js
import express from 'express';
import { createServer } from 'http';
import { Server as SocketIOServer } from 'socket.io';
import cors from 'cors';
import dotenv from 'dotenv';
import { createClient } from 'redis';
import { GoogleGenAI } from '@google/genai';

dotenv.config();

const app = express();
const httpServer = createServer(app);
const allowedOrigin = (origin, cb) => {
  if (!origin || /^https?:\/\/localhost(:\d+)?$/.test(origin)) {
    cb(null, true);
  } else {
    cb(new Error('CORS not allowed'), false);
  }
};

const io = new SocketIOServer(httpServer, {
  cors: {
    origin: allowedOrigin,
    methods: ['GET', 'POST'],
    credentials: true,
  },
  transports: ['websocket', 'polling'],
});

// Middleware
app.use(cors({ origin: allowedOrigin, credentials: true }));
app.use(express.json({ limit: '50mb' }));

// In-memory storage
const activeSessions = new Map();
const canvasData = new Map();

// Redis (optional, for scaling)
let redis = null;
const initRedis = async () => {
  try {
    redis = createClient({
      url: process.env.REDIS_URL || 'redis://localhost:6379',
      socket: {
        reconnectStrategy: (retries) => {
          if (retries > 10) return new Error('Max retries reached');
          return Math.min(retries * 100, 3000);
        }
      }
    });
    redis.on('error', (err) => {
      console.log('⚠️  Redis connection error:', err.message);
      redis = null;
    });
    await redis.connect();
    console.log('✅ Redis connected');
  } catch (err) {
    console.log('⚠️  Redis not available - running in memory mode');
    redis = null;
  }
};
initRedis();

// ============================================
// Gemini AI Client
// ============================================
let genai = null;
const GEMINI_KEY = process.env.GEMINI_API_KEY;
if (GEMINI_KEY && GEMINI_KEY !== 'your-gemini-api-key-here') {
  genai = new GoogleGenAI({ apiKey: GEMINI_KEY });
  console.log('✅ Gemini AI initialized');
} else {
  console.log('⚠️  GEMINI_API_KEY not set — AI features disabled');
}

// ============================================
// REST ENDPOINTS
// ============================================

app.get('/health', (req, res) => {
  const freepikReady = !!(process.env.FREEPIK_API_KEY && process.env.FREEPIK_API_KEY !== 'your-freepik-api-key-here');
  res.json({
    status: 'OK',
    timestamp: new Date().toISOString(),
    redis: redis ? 'connected' : 'offline',
    ai: freepikReady ? 'freepik-mystic' : (genai ? 'gemini-fallback' : 'disabled'),
  });
});

app.get('/api/canvas/:id', (req, res) => {
  const { id } = req.params;
  const canvas = canvasData.get(id) || { strokes: [], elements: [] };
  res.json({ success: true, canvas });
});

// ============================================
// IMAGE GENERATION — Freepik Mystic (primary) + Gemini (fallback)
// ============================================

/**
 * Poll Freepik task until COMPLETED or FAILED (max ~60s)
 */
async function pollFreepikTask(taskId, apiKey, maxWaitMs = 60000) {
  const interval = 2500;
  const deadline = Date.now() + maxWaitMs;
  while (Date.now() < deadline) {
    await new Promise(r => setTimeout(r, interval));
    const r = await fetch(`https://api.freepik.com/v1/ai/mystic/${taskId}`, {
      headers: { 'x-freepik-api-key': apiKey, Accept: 'application/json' },
    });
    const body = await r.json();
    const status = body?.data?.status;
    console.log(`  [Freepik] Task ${taskId} → ${status}`);
    if (status === 'COMPLETED') {
      const urls = body?.data?.generated || [];
      return urls[0] || null;
    }
    if (status === 'FAILED') throw new Error('Freepik task FAILED');
  }
  throw new Error('Freepik task timed out after 60s');
}

/**
 * Generate image via Freepik Mystic API
 */
async function generateWithFreepik(prompt) {
  const FREEPIK_KEY = process.env.FREEPIK_API_KEY;
  if (!FREEPIK_KEY || FREEPIK_KEY === 'your-freepik-api-key-here') {
    throw new Error('FREEPIK_API_KEY not configured');
  }

  console.log(`  → [Freepik Mystic] Posting task...`);
  const resp = await fetch('https://api.freepik.com/v1/ai/mystic', {
    method: 'POST',
    headers: {
      'x-freepik-api-key': FREEPIK_KEY,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
    body: JSON.stringify({
      prompt,
      model: 'realism',
      resolution: '2k',
      aspect_ratio: 'square_1_1',
      creative_detailing: 33,
    }),
  });

  if (!resp.ok) {
    const errBody = await resp.text();
    throw new Error(`Freepik API ${resp.status}: ${errBody}`);
  }

  const data = await resp.json();
  const taskId = data?.data?.task_id;
  if (!taskId) throw new Error('Freepik returned no task_id');

  console.log(`  [Freepik] Task created: ${taskId}`);
  const imageUrl = await pollFreepikTask(taskId, FREEPIK_KEY);
  if (!imageUrl) throw new Error('Freepik completed but returned no image URL');
  return imageUrl;
}

/**
 * Generate image via Gemini (fallback)
 */
async function generateWithGemini(prompt, referenceImage) {
  if (!genai) throw new Error('Gemini not initialized');

  const models = [
    'gemini-2.0-flash-preview-image-generation',
    'gemini-2.0-flash-exp',
  ];

  for (const modelName of models) {
    try {
      console.log(`  → [Gemini] Trying model: ${modelName}`);
      let contents;
      if (referenceImage) {
        const base64Data = referenceImage.replace(/^data:image\/\w+;base64,/, '');
        contents = [
          { text: `Create a high-quality artistic image. Instruction: ${prompt}. Make it seamless, professional, and visually stunning.` },
          { inlineData: { mimeType: 'image/png', data: base64Data } },
        ];
      } else {
        contents = `Create a high-quality artistic image: ${prompt}. Visually stunning, professional quality.`;
      }

      const response = await genai.models.generateContent({
        model: modelName,
        contents,
        config: { responseModalities: ['TEXT', 'IMAGE'] },
      });

      const parts = response.candidates?.[0]?.content?.parts;
      const imagePart = parts?.find(p => p.inlineData?.mimeType?.startsWith('image/'));
      if (imagePart?.inlineData?.data) {
        const dataUrl = `data:${imagePart.inlineData.mimeType};base64,${imagePart.inlineData.data}`;
        return { dataUrl, model: modelName };
      }
      console.log(`  ⚠️ Model ${modelName} returned no image, trying next...`);
    } catch (err) {
      console.error(`  ❌ Gemini model ${modelName} failed: ${err?.message}`);
    }
  }
  throw new Error('All Gemini models failed to return an image');
}

// ─── Main endpoint ───
app.post('/api/generate-image', async (req, res) => {
  const { prompt, referenceImage } = req.body;
  if (!prompt) return res.status(400).json({ error: 'prompt is required' });

  console.log(`\n🎨 Generating image for: "${prompt}"`);

  // 1️⃣ Try Freepik Mystic (primary — best quality, async polling)
  try {
    const imageUrl = await generateWithFreepik(prompt);
    console.log(`✅ Image ready via Freepik Mystic`);
    return res.json({ success: true, imageUrl, model: 'freepik-mystic' });
  } catch (fpErr) {
    console.warn(`  ⚠️ Freepik failed: ${fpErr.message}. Falling back to Gemini...`);
  }

  // 2️⃣ Try Gemini (fallback)
  if (!genai) {
    return res.status(500).json({
      error: 'No image generator available. Add FREEPIK_API_KEY or GEMINI_API_KEY to backend/.env',
    });
  }
  try {
    const { dataUrl, model } = await generateWithGemini(prompt, referenceImage);
    console.log(`✅ Image ready via Gemini (${model})`);
    return res.json({ success: true, imageUrl: dataUrl, model });
  } catch (gemErr) {
    console.error(`  ❌ Gemini fallback failed: ${gemErr.message}`);
    return res.status(500).json({
      error: `Image generation failed. Freepik: not configured or errored. Gemini: ${gemErr.message}`,
    });
  }
});

app.post('/api/canvas', (req, res) => {
  const { name } = req.body;
  const canvasId = `canvas_${Date.now()}`;
  canvasData.set(canvasId, {
    id: canvasId,
    name: name || 'Untitled Canvas',
    strokes: [],
    elements: [],
    createdAt: new Date(),
    updatedAt: new Date(),
  });
  res.json({ success: true, canvas: canvasData.get(canvasId) });
});

// ============================================
// SOCKET.IO EVENTS
// ============================================

io.on('connection', (socket) => {
  console.log(`✅ User connected: ${socket.id}`);

  socket.on('join_canvas', (data) => {
    const { canvasId, userId, username, color } = data;
    socket.join(`canvas:${canvasId}`);

    const session = { userId, socketId: socket.id, username, color, cursorX: 0, cursorY: 0 };
    activeSessions.set(socket.id, { ...session, canvasId });

    const users = Array.from(activeSessions.values())
      .filter((s) => s.canvasId === canvasId)
      .map(({ canvasId: _, ...rest }) => rest);

    socket.emit('active_users', users);
    socket.to(`canvas:${canvasId}`).emit('user_joined', session);
    console.log(`👤 ${username} joined canvas ${canvasId} (Total: ${users.length})`);
  });

  socket.on('draw_stroke', (data) => {
    const { canvasId, stroke } = data;
    if (!canvasData.has(canvasId)) {
      canvasData.set(canvasId, { strokes: [], elements: [] });
    }
    canvasData.get(canvasId).strokes.push(stroke);
    socket.to(`canvas:${canvasId}`).emit('draw_stroke', stroke);
  });

  // Element sync for object model
  socket.on('element_add', (data) => {
    const { canvasId, element } = data;
    if (!canvasData.has(canvasId)) {
      canvasData.set(canvasId, { strokes: [], elements: [] });
    }
    canvasData.get(canvasId).elements.push(element);
    socket.to(`canvas:${canvasId}`).emit('element_add', element);
  });

  socket.on('element_update', (data) => {
    const { canvasId, element } = data;
    const canvas = canvasData.get(canvasId);
    if (canvas) {
      const idx = canvas.elements.findIndex(e => e.id === element.id);
      if (idx >= 0) canvas.elements[idx] = element;
    }
    socket.to(`canvas:${canvasId}`).emit('element_update', element);
  });

  socket.on('element_delete', (data) => {
    const { canvasId, elementId } = data;
    const canvas = canvasData.get(canvasId);
    if (canvas) {
      canvas.elements = canvas.elements.filter(e => e.id !== elementId);
    }
    socket.to(`canvas:${canvasId}`).emit('element_delete', { elementId });
  });

  socket.on('cursor_move', (data) => {
    const { canvasId, x, y } = data;
    const session = activeSessions.get(socket.id);
    if (session) {
      socket.to(`canvas:${canvasId}`).emit('cursor_move', { userId: session.userId, x, y });
    }
  });

  socket.on('leave_canvas', (data) => {
    const { canvasId } = data;
    const session = activeSessions.get(socket.id);
    if (session) {
      socket.leave(`canvas:${canvasId}`);
      socket.to(`canvas:${canvasId}`).emit('user_left', { userId: session.userId });
      activeSessions.delete(socket.id);
      console.log(`❌ ${session.username} left canvas ${canvasId}`);
    }
  });

  socket.on('disconnect', () => {
    const session = activeSessions.get(socket.id);
    if (session) {
      const { canvasId } = session;
      socket.to(`canvas:${canvasId}`).emit('user_left', { userId: session.userId });
      activeSessions.delete(socket.id);
      console.log(`✗ ${session.username} disconnected`);
    }
  });
});

// ============================================
// START SERVER
// ============================================

const PORT = process.env.PORT || 4001;

httpServer.listen(PORT, () => {
  console.log(`\n${'='.repeat(60)}`);
  console.log(`🚀 UNIDRAW Backend Server`);
  console.log(`${'='.repeat(60)}`);
  console.log(`📍 Server:    http://localhost:${PORT}`);
  console.log(`🌐 WebSocket: ws://localhost:${PORT}`);
  console.log(`🔌 CORS:      ${process.env.FRONTEND_URL}`);
  console.log(`📊 Env:       ${process.env.NODE_ENV}`);
  console.log(`🤖 AI:        ${genai ? 'Gemini ✅' : 'Disabled ⚠️'}`);
  console.log(`${'='.repeat(60)}\n`);
});

export { app, io };
